### PR TITLE
HmdaFileParser Serialization

### LINKED ISSUE #833: HmdaFileParser serializer
The default implementation is using `JavaSerializer` by default, which is not recommended due to performance issues

### PR DESCRIPTION
**Pull request by schbetsy**
Closes #833

### CODE REVIEW

File: persistence-model/src/main/resources/application-dev.conf [hunk 1]
```diff
@@ -4,13 +4,26 @@ akka {
...
+      parser = "hmda.persistence.serialization.parser.HmdaFileParserProtobufSerializer"
+      larValidated = "hmda.persistence.serialization.validation.LarValidatedProtobufSerializer"
+      upload = "hmda.persistence.serialization.upload.UploadProtobufSerializer"
     }
     serialization-bindings {
       "hmda.persistence.messages.events.institutions.InstitutionEvents$InstitutionCreated" = institution
       "hmda.persistence.messages.events.institutions.InstitutionEvents$InstitutionModified" = institution
       "hmda.persistence.messages.events.institutions.FilingEvents$FilingCreated" = filing
       "hmda.persistence.messages.events.institutions.FilingEvents$FilingStatusUpdated" = filing
+      "hmda.persistence.messages.events.institutions.SubmissionEvents$SubmissionCreated" = submission
```
Comment (schbetsy) [inline persistence-model/src/main/resources/application-dev.conf]: I wasn't sure about this. I updated `application-dev.conf` to have the same serialization config as `application.conf`. Is that the right way to do this?

Comment (jmarin) [inline persistence-model/src/main/resources/application-dev.conf]: Yes, it's needed in both places. `application-dev.conf` uses leveldb in `demo` mode, while `application.conf` is meant for production and uses `Cassandra`. To me it makes sense to have both using the same serialization format for consistency

File: persistence-model/src/main/scala/hmda/persistence/messages/events/processing/HmdaFileParserEvents.scala [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package hmda.persistence.messages.events.processing
+
+import hmda.model.fi.lar.LoanApplicationRegister
+import hmda.model.fi.ts.TransmittalSheet
+import hmda.model.parser.LarParsingError
+import hmda.persistence.messages.CommonMessages.Event
+
+object HmdaFileParserEvents {
+  trait HmdaFileParserEvent extends Event
```
Comment (nickgrippin) [inline persistence-model/src/main/scala/hmda/persistence/messages/events/processing/HmdaFileParserEvents.scala]: Is this trait needed here?

Comment (schbetsy) [inline persistence-model/src/main/scala/hmda/persistence/messages/events/processing/HmdaFileParserEvents.scala]: @jmarin I saw this pattern in several of your serializer PRs, so I followed it. What's the benefit of using this specialized event trait?

Comment (jmarin) [inline persistence-model/src/main/scala/hmda/persistence/messages/events/processing/HmdaFileParserEvents.scala]: Better type safety. They all inherit from `Event`, which is too generic. One use case might be when reading the journal, maybe I only care about a certain "family" of events.

Comment (nickgrippin) [inline persistence-model/src/main/scala/hmda/persistence/messages/events/processing/HmdaFileParserEvents.scala]: Cool, that's good to know 👍

File: persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufConverterSpec.scala [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import hmda.persistence.serialization.parser.HmdaFileParserProtobufConverter._
+import org.scalacheck.Gen
+import org.scalatest.{ MustMatchers, PropSpec }
+import org.scalatest.prop.PropertyChecks
+
+class HmdaFileParserProtobufConverterSpec extends PropSpec with PropertyChecks with MustMatchers
+    with TsGenerators with LarGenerators {
+
+  def listOfStringsGen: Gen[List[String]] = Gen.listOf(Gen.alphaStr)
+  def intGen: Gen[Int] = Gen.choose(0, 100000)
```
Comment (jmarin) [inline persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufConverterSpec.scala]: I don't think the types are needed here

File: persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufConverterSpec.scala [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  property("TsParsed must convert to protobuf and back") {
+    forAll(tsGen) { ts =>
+      val message = TsParsed(ts)
+      val protobuf = tsParsedToProtobuf(message).toByteArray
+      tsParsedFromProtobuf(TsParsedMessage.parseFrom(protobuf)) mustBe message
+    }
+  }
+
+  property("TsParsedErrors must convert to protobuf and back") {
+    forAll(listOfStringsGen) { errs: List[String] =>
```
Comment (jmarin) [inline persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufConverterSpec.scala]: The type is not needed here

File: persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufConverterSpec.scala [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  property("LarParsed must convert to protobuf and back") {
+    forAll(larGen) { lar =>
+      val message = LarParsed(lar)
+      val protobuf = larParsedToProtobuf(message).toByteArray
+      larParsedFromProtobuf(LarParsedMessage.parseFrom(protobuf)) mustBe message
+    }
+  }
+
+  property("LarParsingError must convert to protobuf and back") {
+    forAll(intGen, listOfStringsGen) { (lineNo: Int, errs: List[String]) =>
```
Comment (jmarin) [inline persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufConverterSpec.scala]: The types are not needed here

File: persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufConverterSpec.scala [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      val message = LarParsingError(lineNo, errs)
+      val protobuf = larParsingErrorToProtobuf(message).toByteArray
+      larParsingErrorFromProtobuf(LarParsingErrorMessage.parseFrom(protobuf)) mustBe message
+    }
+  }
+
+  property("LarParsedErrors must convert to protobuf and back") {
+    forAll(intGen, listOfStringsGen) { (lineNo: Int, errs: List[String]) =>
```
Comment (jmarin) [inline persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufConverterSpec.scala]: The types are not needed here

File: persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufSerializerSpec.scala [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import org.scalatest.{ MustMatchers, PropSpec }
+import org.scalatest.prop.PropertyChecks
+
+class HmdaFileParserProtobufSerializerSpec extends PropSpec with PropertyChecks
+    with MustMatchers with LarGenerators with TsGenerators {
+
+  val serializer = new HmdaFileParserProtobufSerializer()
+
+  def listOfStringsGen: Gen[List[String]] = Gen.listOf(Gen.alphaStr)
+  def intGen: Gen[Int] = Gen.choose(0, 100000)
```
Comment (jmarin) [inline persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufSerializerSpec.scala]: Same comment as above, I don't think you need to explicitly write the types here

File: persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufSerializerSpec.scala [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  }
+  property("LarParsed messages must serialize to binary and back") {
+    forAll(larGen) { lar =>
+      val message = LarParsed(lar)
+      val bytes = serializer.toBinary(message)
+      serializer.fromBinary(bytes, serializer.LarParsedManifest) mustBe message
+    }
+  }
+  property("LarParsedErrors messages must serialize to binary and back") {
+    forAll(intGen, listOfStringsGen) { (lineNo: Int, errs: List[String]) =>
```
Comment (jmarin) [inline persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufSerializerSpec.scala]: Types not needed here

File: persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufSerializerSpec.scala (hunk 7 above)
Comment (schbetsy) [inline persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufSerializerSpec.scala]: I'd like to leave these types here. We've discussed before that `def`s should always be typed, even if the type is obvious.

File: persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufConverterSpec.scala (hunk 4 above)
Comment (schbetsy) [inline persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufConverterSpec.scala]: I'm confused. Is it important to _not_ use the type? If so, why?

Comment (jmarin) [inline persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufConverterSpec.scala]: It's important to be consistent. Most of our code base does not include types in these property based tests (this is not by accident; most code examples you will find out there look like that). And in your own files within this PR you use it sometimes, but not others, I find that confusing. The type here is self explanatory, it doesn't help with readability. I _know_ it's a list of strings.

Comment (schbetsy) [inline persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufConverterSpec.scala]: Okay.

File: persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufConverterSpec.scala (hunk 3 above)
Comment (schbetsy) [inline persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufConverterSpec.scala]: For me, these types greatly improve the readability of the code. Also, see my comment below.

File: persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufSerializerSpec.scala (hunk 7 above)
Comment (jmarin) [inline persistence-model/src/test/scala/hmda/persistence/serialization/parser/HmdaFileParserProtobufSerializerSpec.scala]: I'm fine with leaving the types in these `defs`

**Pull request merged**